### ISSUE TITLE
Too many open files

### ISSUE THREAD

**Issue by kmavrommatis** [ohsu-comp-bio/funnel#614]
I am using a directory as one of the inputs for the task. This directory contains many files and funnel errors out with 'too many open files'.

For the time I am bypassing this issue by creating a tarball with the contents of the directory and untar/unzip it as the first step of the task, but it would be nice to be able to work with the directory data directly

Thanks in advance for your help

**Comment by adamstruck** [comment]
Can you post the specific message? 

Without more information my best guess is that you need to adjust the limits on your host machine as the defaults are generally fairly low for this sort of application. See https://easyengine.io/tutorials/linux/increase-open-files-limit/

**Comment by kmavrommatis** [comment]
Thanks,
I will try to increase the number of allowed open files.

More details:
I have a directory that contains several 100s files. Somewhere near the 200th I get this error:

```
download failed
timestamp | 2019-06-20T03:57:48.973850044Z
task_attempt | 0
executor_index | 0
url | s3://s3.us-east-1.amazonaws.com/celgene-rnd-riku-researchanalytics/data/Genomes/Homo_sapiens/GRCh38.1k/snpEff/GRCh38.86/regulation_CTCF-K562_enriched_sites.bin
error | amazonS3: creating file: open /opt/funnel-work-dir/bk5g7cm4qlf000bl3ng0/celgene-rnd-riku-researchanalytics/data/Genomes/Homo_sapiens/GRCh38.1k/snpEff/GRCh38.86/regulation_CTCF-K562_enriched_sites.bin: too many open files

```

As a general comment, shouldn't the downloader be able to handle arbitrary number of files ? Especially if they are part of a directory?
Is there any reason that keeps the files open after it downloads them?
Thanks in advance for your help

**Comment by adamstruck** [comment]
To answer your questions:

> shouldn't the downloader be able to handle arbitrary number of files ? 

Funnel will download/upload an arbitrary number of inputs or outputs.  Currently all downloads/uploads are started in parallel. Maybe I should put in a configurable limit for how many parallel processes are started at once. 

> Is there any reason that keeps the files open after it downloads them?

File handles are open for the duration of the download and then closed. 


200 input files isn't many, so I am surprised you are hitting the system limit. Does this error occur when running a single task or only when you are running a batch of them?

If it is only in the latter case, I am guessing AWS Batch is packing multiple jobs onto the same EC2 instance and thats why the limits are being hit. You may consider limiting the sizes of instances your compute environment is allowed to spawn to help control how many funnel tasks will be running concurrently on a given instance.

**Comment by kmavrommatis** [comment]
>Maybe I should put in a configurable limit for how many parallel processes are started at once.

That sounds like a good idea.

> I am guessing AWS Batch is packing multiple jobs onto the same EC2 instance

That could very well be the case.  In some other cases the process hit the 'too many open files' when many more files were downloaded. I will investigate a bit more and post an update if I find something related.

**Comment by adamstruck** [comment]
Addressed by #619.

**Comment by adamstruck** [comment]
@kmavrommatis did the patch in #619 fix your issue?
